src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -605,18 +715,27 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
             Compilation currentCompilation,
             ArrayBuilder<TypeParameterDiagnosticInfo> diagnosticsBuilder,
             ref ArrayBuilder<TypeParameterDiagnosticInfo> useSiteDiagnosticsBuilder,
+            ref ArrayBuilder<DiagnosticInfo> tupleDiagnosticsBuilder,
             BitVector skipParameters = default(BitVector))
         {
-            return CheckConstraints(
-                method,
-                conversions,
-                method.TypeSubstitution,
-                ((MethodSymbol)method.OriginalDefinition).TypeParameters,
-                method.TypeArguments,
-                currentCompilation,
-                diagnosticsBuilder,
-                ref useSiteDiagnosticsBuilder,
-                skipParameters);
+            bool result = CheckConstraints(
+                            method,
+                            conversions,
+                            method.TypeSubstitution,
+                            ((MethodSymbol)method.OriginalDefinition).TypeParameters,
+                            method.TypeArguments,
+                            currentCompilation,
+                            diagnosticsBuilder,
+                            ref useSiteDiagnosticsBuilder,
+                            skipParameters);
+
+            foreach (var parameter in method.Parameters)
+            {
+                result &= CheckTupleNamesAgainstBase(parameter.Type, ref tupleDiagnosticsBuilder);
```

Comment:
Makes sense. I'm adding that. There are two questions I'd like to discuss with you tomorrow. Visiting all constituent types and repeating on all bases seems expensive.

1. I think we should avoid check types from metadata. Would that make sense?
2. I'm thinking to add a cache so that types don't get checked twice in a single call to `CheckConstraints`. Does that seem overkill?